pkg/controllers/ragengine_controller.go

```diff
@@ -33,14 +54,381 @@ func NewRAGEngineReconciler(client client.Client, scheme *runtime.Scheme, log lo
+			continue
+		}
+
+		// skip nodes that are not ready
+		_, statusRunning := lo.Find(nodeObj.Status.Conditions, func(condition corev1.NodeCondition) bool {
+			return condition.Type == corev1.NodeReady && condition.Status == corev1.ConditionTrue
+		})
+		if !statusRunning {
+			continue
+		}
+
+		// match the preferred node
+		if preferredNodeSet.Has(nodeObj.Name) {
+			qualifiedNodes = append(qualifiedNodes, lo.ToPtr(nodeObj))
+			continue
+		}
+
+		// match the instanceType
+		if nodeObj.Labels[corev1.LabelInstanceTypeStable] == ragEngineObj.Spec.Compute.InstanceType {
+			qualifiedNodes = append(qualifiedNodes, lo.ToPtr(nodeObj))
+		}
+	}
+
+	return qualifiedNodes, nil
+}
+
+// createAndValidateNode creates a new node and validates status.
+func (c *RAGEngineReconciler) createAndValidateNode(ctx context.Context, ragEngineObj *kaitov1alpha1.RAGEngine) (*corev1.Node, error) {
+	var nodeOSDiskSize string
+
+	if nodeOSDiskSize == "" {
+		nodeOSDiskSize = "0" // The default OS size is used
+	}
+
+	if featuregates.FeatureGates[consts.FeatureFlagKarpenter] {
+		return c.CreateNodeClaim(ctx, ragEngineObj, nodeOSDiskSize)
+	} else {
+		return c.CreateMachine(ctx, ragEngineObj, nodeOSDiskSize)
+	}
+}
+
+func (c *RAGEngineReconciler) CreateMachine(ctx context.Context, ragEngineObj *kaitov1alpha1.RAGEngine, nodeOSDiskSize string) (*corev1.Node, error) {
+RetryWithDifferentName:
+	newMachine := machine.GenerateMachineManifest(ctx, nodeOSDiskSize, ragEngineObj)
+
+	if err := machine.CreateMachine(ctx, newMachine, c.Client); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			klog.InfoS("A machine exists with the same name, retry with a different name", "machine", klog.KObj(newMachine))
+			goto RetryWithDifferentName
+		} else {
+
+			klog.ErrorS(err, "failed to create machine", "machine", newMachine.Name)
+			if updateErr := c.updateStatusConditionIfNotMatch(ctx, ragEngineObj, kaitov1alpha1.ConditionTypeMachineStatus, metav1.ConditionFalse,
+				"machineFailedCreation", err.Error()); updateErr != nil {
+				klog.ErrorS(updateErr, "failed to update ragengine status", "ragengine", klog.KObj(ragEngineObj))
+				return nil, updateErr
+			}
+			return nil, err
+		}
+	}
+
+	// check machine status until it is ready
+	err := machine.CheckMachineStatus(ctx, newMachine, c.Client)
+	if err != nil {
+		if updateErr := c.updateStatusConditionIfNotMatch(ctx, ragEngineObj, kaitov1alpha1.ConditionTypeMachineStatus, metav1.ConditionFalse,
+			"checkMachineStatusFailed", err.Error()); updateErr != nil {
+			klog.ErrorS(updateErr, "failed to update ragengine status", "ragengine", klog.KObj(ragEngineObj))
+			return nil, updateErr
+		}
+		return nil, err
+	}
+
+	// get the node object from the machine status nodeName.
+	return resources.GetNode(ctx, newMachine.Status.NodeName, c.Client)
+}
+
+func (c *RAGEngineReconciler) CreateNodeClaim(ctx context.Context, ragEngineObj *kaitov1alpha1.RAGEngine, nodeOSDiskSize string) (*corev1.Node, error) {
```

Comment:
same logic to https://github.com/Azure/kaito/blob/2ec45ac3141e955bcf37af78ba5f76f5f7e65cdd/pkg/controllers/workspace_controller.go#L577